### PR TITLE
Add `verifyContract` address normalization

### PR DESCRIPTION
**Pull request by OGPoyraz**
## Explanation

This PR aims to add address normalization for `EIP712Domain.verifyContract`

## References

Please see https://github.com/MetaMask/MetaMask-planning/issues/2229 

### Blocked by
- [x] #308

### CODE REVIEW

File: src/utils/normalizer.test.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  it('should normalize verifyingContract address in domain', () => {
+    const normalizedData = parseNormalizerResult(messageData);
+    expect(normalizedData.domain.verifyingContract).toBe(
+      '0xae7ab96520de3a18e5e111b5eaab095312d7fe84',
+    );
+  });
+
+  it('should handle non-hexadecimal verifyingContract address by normalizing it', () => {
```
Comment (matthewwalsh0) [inline src/utils/normalizer.test.ts:75]: How is this different to the above test as both use a decimal input?

File: src/utils/normalizer.test.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
+import { normalizeEIP712TypedMessageData } from './normalizer';
+
+describe('normalizeEIP712TypedMessageData', () => {
+  const messageData = {
```
Comment (matthewwalsh0) [inline src/utils/normalizer.test.ts:4]: Should we instead use a file level constant with uppercase and underscores?

Comment (OGPoyraz) [reply src/utils/normalizer.test.ts]: Done

File: src/utils/normalizer.test.ts (hunk 1 above)
Comment (OGPoyraz) [reply src/utils/normalizer.test.ts]: Removed

[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.test.ts]: Done

File: src/utils/normalizer.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * @param address - The address to normalize.
+ * @returns The normalized address.
+ */
+function normalizeAddress(address: string): Hex {
+  if (isValidHexAddress(address as Hex)) {
+    return address as Hex;
+  }
+
+  // Check if the address is in octal format, convert to hexadecimal
+  if (address.startsWith('0o')) {
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: Good call, I've added as an individual function now

File: src/utils/normalizer.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+/**
+ * Normalizes the address to a hexadecimal format
+ *
+ * @param address - The address to normalize.
+ * @returns The normalized address.
+ */
+function normalizeAddress(address: string): Hex {
+  if (isValidHexAddress(address as Hex)) {
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: That make sense, done

File: src/utils/normalizer.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  return parsedData;
+}
+
+/**
+ * Normalizes the address to a hexadecimal format
+ *
+ * @param address - The address to normalize.
+ * @returns The normalized address.
+ */
+function normalizeAddress(address: string): Hex {
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: Done

File: src/utils/normalizer.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * Parses the messageData to obtain the data object for EIP712 normalization
+ *
+ * @param data - The messageData to parse.
+ * @returns The data object for EIP712 normalization.
+ */
+function parseMessageDataForEIP712Normalization(data: string) {
+  let parsedData = data;
+  if (typeof data === 'string') {
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: Done

File: src/utils/normalizer.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    try {
+      parsedData = JSON.parse(data);
+    } catch (e) {
+      throw new Error(
+        `Invalid message "data" for normalization. data: ${data}`,
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: Done

File: src/utils/normalizer.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+ * @returns The normalized messageData.
+ */
+export function normalizeEIP712TypedMessageData(messageData: string) {
+  const data = parseMessageDataForEIP712Normalization(
+    messageData,
+  ) as unknown as SignTypedMessageDataV3V4;
+
+  const { verifyingContract } = data.domain;
+
+  if (verifyingContract) {
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: Make sense, done

File: src/utils/normalizer.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * @param messageData - The messageData to normalize.
+ * @returns The normalized messageData.
+ */
+export function normalizeEIP712TypedMessageData(messageData: string) {
+  const data = parseMessageDataForEIP712Normalization(
+    messageData,
+  ) as unknown as SignTypedMessageDataV3V4;
+
+  const { verifyingContract } = data.domain;
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: Done

File: src/utils/normalizer.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  message: unknown;
+};
+
+/**
+ * Normalizes the messageData for the eth_signTypedData
+ *
+ * @param messageData - The messageData to normalize.
+ * @returns The normalized messageData.
+ */
+export function normalizeEIP712TypedMessageData(messageData: string) {
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts]: Done

File: src/utils/normalizer.ts
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/utils/normalizer.ts:1]: Done

File: src/wallet.ts [hunk 11]
```diff
@@ -276,7 +277,12 @@ WalletMiddlewareOptions): JsonRpcMiddleware<any, Block> {
     const params = req.params as [string, string];
 
     const address = await validateAndNormalizeKeyholder(params[0], req);
-    const message = params[1];
+    let message = params[1];
+    try {
+      message = normalizeEIP712TypedMessageData(message);
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply src/wallet.ts]: Done

**Comment by OGPoyraz** [comment]
@metamaskbot publish-preview

File: src/utils/normalize.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  try {
+    data = parseTypedMessage(
+      messageData,
+    ) as unknown as SignTypedMessageDataV3V4;
+  } catch (e) {
+    // Ignore normalization errors and pass the message as is
+    return messageData;
+  }
+
+  const { verifyingContract } = data.domain ?? {};
```
Comment (digiwand) [inline src/utils/normalize.ts:33]: this might throw a TypeError if the domain isn't an object as we expect here

Comment (OGPoyraz) [reply src/utils/normalize.ts:33]: Good catch @digiwand but destructing will only throw an error if the value is `null` or `undefined`

Comment (digiwand) [reply src/utils/normalize.ts:33]: Oh good point 👍🏼  domain part lgtm!

**Review by digiwand — DISMISSED** [review]

File: src/wallet.ts [hunk 13]
```diff
@@ -276,7 +277,7 @@ WalletMiddlewareOptions): JsonRpcMiddleware<any, Block> {
     const params = req.params as [string, string];
 
     const address = await validateAndNormalizeKeyholder(params[0], req);
-    const message = params[1];
+    const message = normalizeTypedMessage(params[1]);
```
Comment (mcmire) [inline src/wallet.ts:280]: What is the material effect of this change for this middleware? I know that we have tests for this function in particular but should we add a test for this in `wallet.test.ts` to ensure that the behavior we want is being applied?

File: src/utils/normalize.ts [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * @param messageData - The messageData to normalize.
+ * @returns The normalized messageData.
+ */
+export function normalizeTypedMessage(messageData: string) {
+  let data;
+  try {
+    data = parseTypedMessage(
+      messageData,
+    ) as unknown as SignTypedMessageDataV3V4;
```
Comment (mcmire) [inline src/utils/normalize.ts:27]: When using a type assertion it's good to know why it's necessary. After looking at `parseTypedMessage` I can see it's because of the `JSON.parse`. What are your thoughts on putting this type assertion in that function instead so it's more associated with the `JSON.parse` visually?

File: src/utils/normalize.ts [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+/**
+ * Parses the messageData to obtain the data object for EIP712 normalization
+ *
+ * @param data - The messageData to parse.
+ * @returns The data object for EIP712 normalization.
+ */
+function parseTypedMessage(data: string) {
+  if (typeof data !== 'string') {
```
Comment (mcmire) [inline src/utils/normalize.ts:51]: Do we have a test for this?

File: src/utils/normalize.ts [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * Normalizes the address to a hexadecimal format
+ *
+ * @param address - The address to normalize.
+ * @returns The normalized address.
+ */
+function normalizeContractAddress(address: string): Hex {
+  const addressHex = address as Hex;
+  if (isValidHexAddress(addressHex)) {
```
Comment (mcmire) [inline src/utils/normalize.ts:70]: It appears that we need to use a type assertion because `isValidHexAddress` takes a `Hex`. However, we should be able to avoid the type assertion by using `isStrictHexString`, which narrows its argument to a `Hex` if given a hex string:

``` suggestion
  if (isStrictHexString(address) && isValidHexAddress(address)) {
```

That means we shouldn't need to use `addressHex` below, or upcast it to a `string`.

File: src/utils/normalize.ts [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  // Check if the address is in octal format, convert to hexadecimal
+  if (addressHex.startsWith('0o')) {
+    // If octal, convert to hexadecimal
+    return octalToHex(addressHex as string);
+  }
+
+  // Check if the address is in decimal format, convert to hexadecimal
+  const parsedAddress = parseInt(addressHex, 10);
```
Comment (mcmire) [inline src/utils/normalize.ts:81]: Do we need to be worried that this will be a large number? I see that you use `BN` below, but you seem to only use it for conversion (and then, only for decimal -> hex and not octal -> hex).

File: src/utils/normalize.ts [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+ */
+function parseTypedMessage(data: string) {
+  if (typeof data !== 'string') {
+    return data;
+  }
+
+  try {
+    return JSON.parse(data);
+  } catch (e) {
+    throw new Error(`Invalid message data for normalization. data: ${data}`);
```
Comment (mcmire) [inline src/utils/normalize.ts:58]: Is there value in throwing a custom error here since we catch it above anyway?

File: src/utils/normalize.ts [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  // Check if the address is in decimal format, convert to hexadecimal
+  const parsedAddress = parseInt(addressHex, 10);
+  if (!isNaN(parsedAddress)) {
+    const hexString = new BN(addressHex.toString(), 10).toString(16);
+    return add0x(hexString);
+  }
+
+  // Returning the original address without normalization
```
Comment (mcmire) [inline src/utils/normalize.ts:87]: Do we have a test for this?

File: src/utils/normalize.ts (hunk 14 above)
Comment (OGPoyraz) [reply src/utils/normalize.ts]: That make sense @mcmire, fixed in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/2ebe6855a76c9953f2054d108b606049a75a233e

File: src/utils/normalize.ts (hunk 15 above)
Comment (OGPoyraz) [reply src/utils/normalize.ts:49]: Done in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/af2a8c25a11e53caed2f62ce00730bc271f3e839

File: src/utils/normalize.ts (hunk 16 above)
Comment (OGPoyraz) [reply src/utils/normalize.ts]: Good call, done in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/a8e8e716d05de8b8d46afff0d397dcda2cf977bc

File: src/utils/normalize.ts (hunk 17 above)
Comment (OGPoyraz) [reply src/utils/normalize.ts]: Done in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/ba7ec73b344b427085f6eabe09045eb824284cc6

File: src/utils/normalize.ts (hunk 18 above)
Comment (OGPoyraz) [reply src/utils/normalize.ts]: Removed in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/4baeb92223cdc969c9ad9af1a9aece9a93c6a3a9

File: src/utils/normalize.ts (hunk 19 above)
Comment (OGPoyraz) [reply src/utils/normalize.ts:82]: Added in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/7c2e14126aa897cdaa9475d53877d5b9afd301c1

File: src/wallet.test.ts [hunk 20]
```diff
@@ -326,6 +331,68 @@ describe('wallet', () => {
...
+      expect(signTypedDataV3Result).toStrictEqual(testMsgSig);
+      expect(witnessedMsgParams).toHaveLength(1);
+      expect(witnessedMsgParams[0]).toMatchObject({
+        from: testAddresses[0],
+        data: expectedStringifiedMessage,
+        version: 'V3',
+        signatureMethod: 'eth_signTypedData_v3',
+      });
+    });
+  });
```
Comment (digiwand) [inline src/wallet.test.ts:394]: [suggestion] could include tests for:
- octal addresses
- inputs that are not parsable

Comment (OGPoyraz) [reply src/wallet.test.ts:394]: As mentioned above (https://github.com/MetaMask/eth-json-rpc-middleware/pull/309#discussion_r1651252283), these test cases are already covered by the normalizer unit tests. I added this single eth_signTypedData_v3 test because it was missing from the wallet tests. Is there any advantage to adding essentially the same tests here as well?

File: src/wallet.ts (hunk 13 above)
Comment (mcmire) [reply src/wallet.ts:280]: Yeah, a simple test is fine, no need to test edge cases. Thanks :)

**Review by mcmire — APPROVED** [review]
I'm happy with this, looks good to me.
